gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageReadChannel.java

```diff
@@ -1060,6 +1060,42 @@ protected InputStream openStream(long bytesToRead) throws IOException {
     }
   }
 
+  private void cacheFooterWithRetries(Storage.Objects.Get getObject, HttpResponse response)
+      throws IOException {
+    for (int retriesCount = 0; retriesCount < maxRetries; retriesCount++) {
+      try {
+        cacheFooter(response);
+        if (retriesCount != 0) {
+          logger.atInfo().log(
+              "Successfully cached footer after %s retries for '%s'", retriesCount, resourceId);
+        }
+        break;
+      } catch (IOException footerException) {
+        logger.atInfo().withCause(footerException).log(
+            "Failed to prefetch footer (retry #%s/%s) for '%s'",
+            retriesCount + 1, maxRetries, resourceId);
+        if (retriesCount == 0) {
+          readBackOff.get().reset();
+        }
+        if (retriesCount == maxRetries) {
+          resetContentChannel();
+          throw footerException;
+        }
+        try {
+          response = getObject.executeMedia();
+          // TODO(b/110832992): validate response range header against expected/request range.
+        } catch (IOException e) {
+          response = handleExecuteMediaException(e);
+        }
+      }
+    }
+    checkState(
```

Comment:
Even if perfecting of footer failed is it necessary to fail the `open` request? We already have got metadata. If there is a read request to fetch footer content it can and will be retried at that time.